Mac MiniBrowser should match Safari's behavior for anchor links to javascript URLs

rreno opened the pull request:
#### 61e2b7c76840679081dd4aa1518d84f006cdf175
<pre>
Mac MiniBrowser should match Safari&apos;s behavior for anchor links to javascript URLs
<a href="https://bugs.webkit.org/show_bug.cgi?id=246950">https://bugs.webkit.org/show_bug.cgi?id=246950</a>
rdar://101502667

Reviewed by Brent Fulgham.

Safari on Mac cancels navigation from an anchor element to a javascript URL when target=_blank.
As it is currently implemented, MiniBrowser will attempt to download what appears to be the result
of executing the javascript. This behavior difference can cause some confusion when debugging tests
or web pages using MiniBrowser.

This change mirrors Safari&apos;s behavior in Mac MiniBrowser by instructing WebKit to cancel the
navigation in the case of a javascript URL.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(isJavaScriptURL):
(-[WK2BrowserWindowController webView:decidePolicyForNavigationAction:decisionHandler:]):

Canonical link: <a href="https://commits.webkit.org/255918@main">https://commits.webkit.org/255918@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0d42f2e1ea2f71cf83baef6d3d22650fe44b140

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3217 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/103657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3228 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31450 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86340 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99684 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80446 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29334 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37839 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35708 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39584 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1923 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41520 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
```diff
@@ -761,9 +766,11 @@ - (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigati
     }
 
-    if (navigationAction._userInitiatedAction && !navigationAction._userInitiatedAction.isConsumed) {
+    NSURL* url = navigationAction.request.URL;
```
Nit: In Objective C code we write this as: `NSURL *url = navigationAction.request.URL`

brentfulgham replied:
Also: Is a similar change needed in WK1BrowserWindowController for the Legacy WebKit code path?

rreno replied:
Ah the pointer goes with the variable instead of the type? Nice.

rreno replied:
I don't think a similar change is needed for WK1. The API is slightly different and we unconditionally tell the PolicyDecisionListener to go ahead with the navigation in MiniBrowser WK1. I also don't think there's going to be confusion in WK1 mode between Safari and MiniBrowser since Safari is WK2 and there's probably many more divergences in behavior between a WK1 and WK2 browser.

brentfulgham reviewed (commented):
r=me

rreno committed:
Mac MiniBrowser should match Safari's behavior for anchor links to javascript URLs
https://bugs.webkit.org/show_bug.cgi?id=246950
rdar://101502667

Reviewed by Brent Fulgham.

Safari on Mac cancels navigation from an anchor element to a javascript URL when target=_blank.
As it is currently implemented, MiniBrowser will attempt to download what appears to be the result
of executing the javascript. This behavior difference can cause some confusion when debugging tests
or web pages using MiniBrowser.

This change mirrors Safari's behavior in Mac MiniBrowser by instructing WebKit to cancel the
navigation in the case of a javascript URL.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(isJavaScriptURL):
(-[WK2BrowserWindowController webView:decidePolicyForNavigationAction:decisionHandler:]):

Canonical link: https://commits.webkit.org/255918@main
